plugins/CoreUpdater/Updater.php

```diff
@@ -124,15 +126,20 @@ public function updatePiwik($https = true)
             throw new UpdaterException($e, $messages);
         }
 
-        $partTwoUrl = Url::getCurrentUrlWithoutQueryString() . Url::getCurrentQueryStringWithParametersModified([
-            'action' => 'oneClickUpdatePartTwo',
-        ]);
+        $validFor10Minutes = time() + (60 * 10);
+        $nonce = Common::generateUniqId();
+        Option::set('NonceOneClickUpdatePartTwo', json_encode(['nonce' => $nonce, 'ttl' => $validFor10Minutes]));
 
-        $response = Http::sendHttpRequest($partTwoUrl, 300);
-        $response = @json_decode($response, $assoc = true);
+        PiwikCache::flushAll();
 
-        if (!empty($response)) {
-            $messages = array_merge($messages, $response);
+        $cliMulti = new CliMulti();
+        $responses = $cliMulti->request(['?module=CoreUpdater&action=oneClickUpdatePartTwo&nonce=' . $nonce]);
+        if (!empty($responses)) {
+            $response = array_shift($responses);
+            $response = @json_decode($response, $assoc = true);
+            if (!empty($response) && is_array($response)) {
+                $messages = array_merge($messages, $response);
```

Comment:
one problem here is that any error during  `oneClickUpdatePartTwo` won't trigger an exception and therefore no error will be set here https://github.com/matomo-org/matomo/blob/3.14.1-b1/plugins/CoreUpdater/Controller.php#L167 and therefore the user won't see any of the messages should something go wrong. This was already a problem with the original solution but don't know if we want to show error messages for this action or simply ignore them as it affects only marketplace plugins? And they could simply update them afterwards

Reply:
Do you mean a fatal error? Or something else? If it's a fatal error, perhaps the safemode hook could be used. Not sure if it's worth the effort though.

Reply:
@diosmosis I meant eg if there's an error like I had it `Got invalid response from API request: https://.... Response was 'curl_exec: SSL certificate problem: unable to get local issuer certificate. Hostname requested was: ...'` or if there's suddenly a bug with the `nonce` and the call is not actually working etc. Before the token_auth was missing so the action for the 2nd step would have always returned an error but the error was never shown.

Reply:
I guess we'd have to handle this when the controller action is called? Ie, "cannot parse json response, response is: curl-exec: ..." or "invalid token, could not complete install". or maybe i'm missing something?

Reply:
Note: I don't know if it's actually that important.